prov/shm/src/smr.h

```diff
@@ -91,36 +91,64 @@ int smr_av_open(struct fid_domain *domain, struct fi_av_attr *attr,
 #define SMR_IOV_LIMIT		4
 
 struct smr_ep_entry {
+	struct dlist_entry	entry;
 	void			*context;
 	fi_addr_t		addr;
+	uint64_t		tag;
+	uint64_t		ignore;
 	struct iovec		iov[SMR_IOV_LIMIT];
 	uint8_t			iov_count;
 	uint8_t			flags;
 	uint8_t			resv[sizeof(size_t) - 2];
 };
 
-OFI_DECLARE_CIRQUE(struct smr_ep_entry, smr_rx_cirq);
-
 struct smr_ep;
 typedef void (*smr_rx_comp_func)(struct smr_ep *ep, void *context,
-		uint64_t flags, size_t len, void *buf, void *addr);
-typedef void (*smr_tx_comp_func)(struct smr_ep *ep, void *context);
+		uint64_t flags, size_t len, void *buf, void *addr,
+		uint64_t tag);
+typedef void (*smr_tx_comp_func)(struct smr_ep *ep, void *context,
+		uint64_t flags);
+
+DECLARE_FREESTACK(struct smr_ep_entry, smr_recv_fs);
+
+struct smr_recv_match_attr {
+	fi_addr_t	addr;
+	uint64_t	tag;
+};
+
+static inline int smr_match_addr(fi_addr_t addr, fi_addr_t match_addr)
+{
+	return (addr == FI_ADDR_UNSPEC) || (match_addr == FI_ADDR_UNSPEC) ||
+		(addr == match_addr);
+}
+
+static inline int smr_match_tag(uint64_t tag, uint64_t ignore, uint64_t match_tag)
+{
+	return ((tag | ignore) == (match_tag | ignore));
+}
+
+struct smr_recv_queue {
+	struct dlist_entry recv_list;
+	dlist_func_t *match_recv;
+};
 
 struct smr_ep {
 	struct util_ep		util_ep;
 	smr_rx_comp_func	rx_comp;
 	smr_tx_comp_func	tx_comp;
 	size_t			tx_size;
-	struct smr_rx_cirq	*rxq; /* protected by rx_cq lock */
+	size_t			rx_size;
 	const char		*name;
 	struct smr_region	*region;
 	uint32_t		msg_id;
+	struct smr_recv_fs	*recv_fs; /* protected by rx_cq lock */
+	struct smr_recv_queue	recv_queue;
+	struct smr_recv_queue	trecv_queue;
```

Comment:
You will need to setup an unexpected message queue.  For tagged messages that don't find a matching receive, the message should be stored as an unexpected message. When a tagged receive is posted, the unexpected queue should be checked for a match prior to posting the message to the general tagged receive queue.  This ensures that an unmatched message doesn't end up blocking the receive side command queue.

The unexpected queue can buffer the entire message, for small transfers, or just record the iov where the data is.  In the first case, the send can be completed as soon as the message is queued in the unexpected queue.  In the latter case, the send cannot be completed until the receive has been posted.

The unexpected queue concept can be used with untagged messages if desired as a form of receive side buffering, unless some other form of flow control is used.